Check if a fatal warning issued in typer is silenced, before converting it into an error

KacperFKorban opened the pull request:
closes lampepfl#17741
closes lampepfl#17735

dwijnand reviewed (changes requested):
I have a number of issues with this.

First the premise: I don't see how this is doing anything before converting it to an error, if anything it looks like it's converting warnings to errors before running through Wconf.

Your test doesn't seem to capture the intent of the original 17741 and the issue 17735 doesn't have a test case either.  Your 17741 seems to be triggering a auto-apply-insertion warning, which is correctly suppressed by `@nowarn`.  So it's nothing to do with `-Wnonunit-statement`, which is also missing from the test case anyways.

KacperFKorban commented:
Converting warnings to errors before checking `-Wconf` was a mistake, I fixed it now and added a test to make sure.

This change is supposed to address the fact that when using `StoreReporter` with `fromTyperState = true`, `issueUnconfigured` was called before `issueIfNotSuppressed`. Because of that, warnings under `-Xfatal-warnings` were converted to errors before any check for `-Wconf` or `@nowarn` was done. (related PR: https://github.com/lampepfl/dotty/pull/13911)

The bug had nothing to do with `-Wnonunit-statement` or `-Wvalue-discard`, so I used a simpler test before. I added the tests for i17735 and i17741 now. (Though in i17741, I used a macro-expanded version)

KacperFKorban commented:
@dwijnand Have you had some time to read my last comment? (It might have gotten buried in your notifications)
